### PR TITLE
Adaptive Transitional marcov chain monte carlo

### PR DESCRIPTION
**Pull request by hvasbath**
Finally to give you guys something back and show my gratitude for putting pymc together:
An initial draft for Adaptive Transitional marcov chain monte carlo.
Which is best suited for high dimensional, non-linear problems where no gradient can be computed.

It cannot be used with the traditional 'sample' function and has its own ATMIP_sample.
So far only for Text backend I havent tried the other backends.
I am of course grateful for any suggestions for improvements and hope you guys like it.

It comes with a test:

``` javascript
import pymc3 as pm
import numpy as num
import ATMCMC as atmcmc
import theano.tensor as tt
from matplotlib import pylab as pl

test_folder = ('/data/SAR_data/Aqaba1995/subsampled/model_result_InSAR_Seis/'
               'ATMIP_TEST')

N_chains = 1000
N_steps = 100
tune_interval = 25
njobs = 1

n = 2

mu1 = num.ones(n) * (1. / 2)
mu2 = -mu1

stdev = 0.1
sigma = num.power(stdev, 2) * num.eye(n)
isigma = num.linalg.inv(sigma)
dsigma = num.linalg.det(sigma)

w1 = stdev
w2 = (1 - stdev)


def two_gaussians(x):
    log_like1 = - 0.5 * n * tt.log(2 * num.pi) \
                - 0.5 * tt.log(dsigma) \
                - 0.5 * (x - mu1).T.dot(isigma).dot(x - mu1)
    log_like2 = - 0.5 * n * tt.log(2 * num.pi) \
                - 0.5 * tt.log(dsigma) \
                - 0.5 * (x - mu2).T.dot(isigma).dot(x - mu2)
    return tt.log(w1 * tt.exp(log_like1) + w2 * tt.exp(log_like2))

with pm.Model() as ATMIP_test:
    X = pm.Uniform('X',
                   shape=n,
                   lower=-5. * num.ones_like(mu1),
                   upper=5. * num.ones_like(mu1),
                   testval=-4. * num.ones_like(mu1),
                   transform=None)
    like = pm.Deterministic('like', two_gaussians(X))
    llk = pm.Potential('like', like)

with ATMIP_test:
    step = atmcmc.ATMCMC(N_chains=N_chains, tune_interval=tune_interval,
                         likelihood_name=ATMIP_test.deterministics[0].name)

trcs = atmcmc.ATMIP_sample(
                        N_steps=N_steps,
                        step=step,
                        njobs=njobs,
                        progressbar=False,
                        trace=test_folder,
                        model=ATMIP_test)

pm.summary(trcs)
Pltr = pm.traceplot(trcs, combined=True)
pl.show(Pltr[0][0])
```

edited test.

### CODE REVIEW

**Comment by fonnesbeck** [comment]
Thanks for this! Do you have a reference handy for this method?

**Comment by hvasbath** [comment]
The code here is actually rather the CATMIP reference, but CATMIP it is very similar to ATMCMC which is the original. The references are written in the code. But here are the links:
ATMCMC
http://ascelibrary.org/doi/abs/10.1061/%28ASCE%290733-9399%282007%29133:7%28816%29
CATMIP
https://gji.oxfordjournals.org/content/194/3/1701.full

File: pymc3/step_methods/ATMCMC.py [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+from pymc3.step_methods.metropolis import MultivariateNormalProposal as MvNPd
+from numpy.random import seed
+from joblib import Parallel, delayed
+
+__all__ = ['ATMCMC', 'ATMIP_sample']
+
+
+class ATMCMC(pm.arraystep.ArrayStepShared):
+    """
+    Transitional Marcov-Chain Monte-Carlo
```
Comment (twiecki) [inline pymc3/step_methods/ATMCMC.py]: Markov

**Comment by hvasbath** [comment]
So finally also my complex calculation is running on that. I guess it will run a week. I am sorry for posting the code too early. Now you have all the mess here as well. But there shouldnt be too many more changes from my side coming at least this frequently.

**Comment by twiecki** [comment]
@eigenblutwurst Can you add an example NB that shows it? I see you posted one above but it doesn't use your PR it seem (`import ATMCMC as atmcmc`).

**Comment by hvasbath** [comment]
It does use exactly this one. So this is the example notebook. If you check "files changed" you see it is stored as ATMCMC.py, unfortunately I named the git branch TMCMC and later added the A to the file. Can I change git branch names? As already mentioned I am really new to using git ;) .

**Comment by hvasbath** [comment]
But nevermind testing it. At the moment it is again crashing for some reason. So please wait until the next commit. Hmpf. From one second to the next it starts crashing wth ...

**Comment by springcoil** [comment]
Thanks so much for doing this. Will be interesting to see this getting resolved.

**Pull request closed without merging**